--- sgmlop/sgmlop.c	2013-03-28 10:43:34.000000000 -0600
+++ sgmlop/sgmlop.c	2013-03-28 11:15:32.000000000 -0600
@@ -116,13 +116,14 @@
 #define PyObject_GC_Del PyObject_Del
 #define PyObject_GC_Track
 #define PyObject_GC_UnTrack
-#define PyGC_HEAD_SIZE 0
 typedef int (*visitproc)(PyObject *, void *);
 typedef int (*traverseproc)(PyObject *, visitproc, void *);
 #endif
 
 #endif
-
+#ifndef PyGC_HEAD_SIZE
+#define PyGC_HEAD_SIZE 0
+#endif
 /* ==================================================================== */
 /* parser data type */
 
@@ -178,7 +179,7 @@
 
 } FastParserObject;
 
-staticforward PyTypeObject FastParser_Type;
+static PyTypeObject FastParser_Type;
 
 /* forward declarations */
 static int fastfeed(
@@ -557,9 +558,8 @@
     return Py_FindMethod(_sgmlop_methods, (PyObject*) self, name);
 }
 
-statichere PyTypeObject FastParser_Type = {
-    PyObject_HEAD_INIT(NULL)
-    0, /* ob_size */
+static PyTypeObject FastParser_Type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
     "FastParser", /* tp_name */
     sizeof(FastParserObject), /* tp_size */
     0, /* tp_itemsize */
@@ -585,8 +585,9 @@
 #endif
 initsgmlop(void)
 {
-    /* Patch object type */
-    FastParser_Type.ob_type = &PyType_Type;
+    /* Initialize object types */
+    if (PyType_Ready(&FastParser_Type) < 0)
+        return;
 
 #if defined (SGMLOP_GC)
     /* enable garbage collection for the parser object */
